### PR TITLE
Extended Staggered Scheme in Multi-component Transport Process

### PR DESCRIPTION
**Pull request by renchao-lu**
This pull request is based upon #2341. With this implementation, staggered scheme is ready for multi-component transport simulation.

### CODE REVIEW

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 1]
```diff
@@ -358,27 +373,329 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
+
+            auto const local_p = Eigen::Map<const NodalVectorType>(
+                &local_x[pressure_index], pressure_size);
+            int const concentration_index = 1 * ShapeFunction::NPOINTS;
+            auto const local_C = Eigen::Map<const NodalVectorType>(
+                &local_x[concentration_index], concentration_size);
+
+            return calculateIntPtDarcyVelocity(t, local_p, local_C, cache);
+        }
+        else  // staggered scheme
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: no else after return...

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 2]
```diff
@@ -358,27 +373,329 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
...
 
-        // Assuming that fluid density always depends on the concentration of
-        // the component placed at the first.
-        auto const concentration_index = 1 * ShapeFunction::NPOINTS;
-        // get local_C and local_p
-        auto const C_nodal_values = Eigen::Map<const NodalVectorType>(
-            &local_x[concentration_index], concentration_size);
-        auto const p_nodal_values = Eigen::Map<const NodalVectorType>(
-            &local_x[pressure_index], pressure_size);
+        if (_coupled_solutions == nullptr)  // monolithic scheme
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: if (_use_monolithic_scheme)

**Comment by renchao-lu** [comment]
Done. Please review.

File: ProcessLib/ComponentTransport/ComponentTransportProcess.h [hunk 3]
```diff
@@ -146,7 +146,7 @@ class ComponentTransportProcess final : public Process
         _local_assemblers;
 
     /// Solutions of the previous time step
-    std::array<std::unique_ptr<GlobalVector>, 2> _xs_previous_timestep;
+    std::vector<std::unique_ptr<GlobalVector>> _xs_previous_timestep;
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportProcess.h]: Is this a bug fix?

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 4]
```diff
@@ -499,6 +499,7 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
         std::vector<double>& /*local_b_data*/,
         LocalCoupledSolutions const& coupled_xs)
     {
+        auto const& transport_process_id = coupled_xs.process_id;
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Is it correct that this change is only relevant in the staggered scheme? If so the variable name should reflect this special use. What do you think about `transport_process_id_staggered`?

File: ProcessLib/ComponentTransport/ComponentTransportProcess.h (hunk 3 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportProcess.h]: Last pull request (#2341) achieved functionality of staggered coupling scheme in single component transport process which has two primary variables, i.e., one for hydraulic pressure and the other for component concentration. Currently, we would like to use staggered scheme for multi-component transport simulation.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h (hunk 4 above)
Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: `transport_process_id` was set as a static constant variable in previous implementation (#2341) where single component transport process is given consideration. This change is to get a particular transport process id from local coupled solutions where solutions of multiple transport processes are stored.

Comment (TomFischer) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: What do you think about `particular_transport_process_id` or `coupled_solutions_transport_process_id`? Please let the reader of your code know that this is not the usual `some_process_id`, but it is used here in the staggered context.

Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Changed to `particular_transport_process_id`.

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 5]
```diff
@@ -80,7 +80,7 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
     // When staggered scheme is adopted, nodal pressure and nodal concentration
     // are accessed by process id.
     static const int hydraulic_process_id = 0;
-    static const int transport_process_id = 1;
+    static const int first_transport_process_id = 1;
```
Comment (TomFischer) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: IMHO, for consistent naming scheme, it would be better to revert this name change.

Comment (renchao-lu) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: `first_transport_process_id` is used for calculating fluid density, with the assumption that fluid density is dependent upon the concentration of the component which is placed uppermost. There is no such an issue in single component transport simulation.

Comment (endJunction) [reply ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Then I recommend to add a comment explaining the variable purpose (if it cannot be explained by variable name).
```c++
    // TODO (renchao-lu): The variable is used in the calculation of the fluid's
    // density and flux indicating the transport process id. For now it is
    // assumed that these quantities depend on the first occurring transport
    // process only. The density and flux calculations have to be extended to
    // all processes.
```
*sic*!

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 6]
```diff
@@ -499,8 +499,9 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
         std::vector<double>& /*local_b_data*/,
         LocalCoupledSolutions const& coupled_xs)
     {
+        auto const& particular_transport_process_id = coupled_xs.process_id;
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: In my opinion, this is the only line (aside the vector and resize) relevant to the described changes.
```c++
auto const& transport_process_id = coupled_xs.process_id;
```
All of the renames can be dropped.

**Pull request merged**